DROOLS-3023: [DMN Designer] Data Types - Warnings - User must be warned when some operations affect other Data Types

karreiro opened the pull request:
See: https://issues.jboss.org/browse/DROOLS-3023

![demo_](https://user-images.githubusercontent.com/1079279/46708451-0e3c1c00-cc15-11e8-9e3c-7a5be64c0e19.gif)

karreiro commented:
Jenkins please retest this.

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/confirmation/DataTypeConfirmation.java:
```diff
@@ -0,0 +1,99 @@
+
+    private boolean hasLostSubDataTypes(final DataType dataType) {
+        final ItemDefinition itemDefinition = itemDefinitionStore.get(dataType.getUUID());
+        return !isStructure(dataType) && itemDefinition.getTypeRef() == null;
```
We have covered this private method by two tests:
- `testIfDataTypeDoesNotHaveLostSubDataTypesWhenCallbackIsExecuted`
- `testIfDataTypeDoesNotHaveLostSubDataTypesWhenCallbackIsNotExecuted`

However in both tests `itemDefinition.getTypeRef() == null` is evaluated to true what makes me ask if this part of condition is really needed or if it is kind of forgotten from early development phase.

karreiro replied:
"Fixed" ;-)

jomarko commented:
The message `All fields that are part of this structure will be removed as a result of this change. Are you sure you want to change this Data Type?` remains shown between two openings of manage custom data type dialog.
- Open **Manage Custom Data Type** dialog
- Click edit next to **Structure**
- Select **Simple Data Type**
- Try to save this data type entry
- Warning appears - **correct**
- Cancel **Manage Custom Data Type** dialog by cross in upper right corner
- Open again **Manage Custom Data Type** dialog
- Warning appears - **incorrect**

jomarko commented:
Probably just my **non-native** English skills. But my understanding of **All Data Types with the "tProfile" type will be removed. Are you sure you want to remove this Data Type?** is that also Data Types that has field of **tProfile** type will be removed however we remove just all occurrences of **tProfile** fields, for me it would make sense to change the message. However again, I am not native speaker, leaving decision up to you.

jomarko commented:
I think I am done, however originally thought there will be warnings also if removed item is set as output data type for some element in model, will be this addressed in separate issues? Should file them or do we have already some?

manstis commented:
Jenkins please retest this.

karreiro commented:
@jomarko, Thank you for the review.

> The message "All fields that are part of this structure will be removed as a result of this change. Are you sure you want to change this Data Type?" is shown even if you change Structure without fields to simple data type, should I file enhancement task for this?

Fixed (see the demo).


> The message All fields that are part of this structure will be removed as a result of this change. Are you sure you want to change this Data Type? remains shown between two openings of manage custom data type dialog.

Fixed.


> Probably just my non-native English skills. But my understanding of All Data Types with the "tProfile" type will be removed. Are you sure you want to remove this Data Type? is that also Data Types that has field of tProfile type will be removed however we remove just all occurrences of tProfile fields, for me it would make sense to change the message. However again, I am not native speaker, leaving decision up to you.

I not a native English speaker too (so I asked some help to Stetson for these messages) 😊 Every row in the Data Type list view is a Data Type, sometimes we do call them "fields". However, I think that in this case would strange a sentence like  **"All fields with the "tProfile" type will be removed. Are you sure you want to remove this Data Type?"**, because sometimes a top-level Data Type can be removed too, in a scenario like this:
```
- tPerson (Structure)
	- age (Number)
	- name (String)
- person (tPerson)
- tCity (Structure)
	- mayor (tPerson)
```
..in the example above, the ~field~ Data Type `mayor` and the Data Type `person` will be removed.


> I think I am done, however originally thought there will be warnings also if removed item is set as output data type for some element in model, will be this addressed in separate issues? Should file them or do we have already some?

We can raise another JIRAs for the scenarios that you pointed. For the DROOLS-3023, we're covering only these two. 

---

@manstis, Thank you for the review.

> Removing nested field from top-level structure did not report it was used elsewhere. Should it alert?

No. In this PR, I'm covering only the two scenarios described in the DROOLS-3023. We could raise a JIRA for the scenario above, but personally, I wouldn't since it's not a _very_ destructive scenario, I think that does not worth to stop the users flow with a warning message in this case. But, again, it's my personal opinion, if you think that it's an important scenario, we can raise a JIRA for that.


> Changing top level structure to a primitive shows the "Are you sure..." warning, but cancelling the warning message does not cancel the edit operation. Should it? I can cancel by then clicking the "x" just wondering whether the User would expect the change to revert too.

Yeah, it was canceling only the save operation. But I think you're right, makes sense to close the edit mode in the cancel the operation. Now it's fixed :-) (see the demo)

---

Demo:
![demo](https://user-images.githubusercontent.com/1079279/46971307-01df1580-d092-11e8-9499-f9e32716809e.gif)

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/messages/DataTypeFlashMessages.java:
```diff
@@ -71,6 +77,25 @@ void highlightDataField(final DataTypeFlashMessage flashMessage) {
+    void registerFlashMessageCallback(final DataTypeFlashMessage flashMessage) {
+        switch (flashMessage.getType()) {
+            case ERROR:
+                // 'Error' FlashMessage does not have callbacks.
```
Hmmm... personally I would prefer to keep it as is 🤔

The callbacks `warningSuccessCallback` and `warningErrorCallback` are used only by buttons in the `warning-container` DIV, so I don't think that an error message should modify them.

In fact, resetting these callbacks would create a bug, because we can show an error message while a warning message is being shown. So, we need to preserve the callbacks to keep both message healthy.

Probably we'll split the `DataTypeFlashMessages` in two singleton classes in the future (..but let's wait for the need of adding success messages in the dialog hehe).

manstis replied:
Thank-you for the explanation.

karreiro commented:
Jenkins please retest this.
